### ISSUE TITLE
[gcp] Node mistakenly marked dead: increase heartbeat timeout?

### ISSUE THREAD

**Issue by ahah-figure** [ray-project/ray#16945]
I’m using ray on a GCP gpu cluster for hyperparameter tuning, training, and prediction. For any of these use cases, about 25% of the time, ray crashes with the following message:

> The node with node id: xxx and ip: xxx has been marked dead because the detector has missed too many heartbeats from it. This can happen when a raylet crashes unexpectedly or has lagging heartbeats.

On nodes that are marked dead, I go to raylet.out, and I see many messages like

> Last resource report was sent 612 ms ago. There might be resource pressure on this node. If resource reports keep lagging, scheduling decisions of other nodes may become stale

and

> Last heartbeat was sent 515 ms ago. There might be resource pressure on this node. If heartbeat keeps lagging, this node can be marked as dead mistakenly

Most warnings are in the 500 ms range. But I do see a few as high as 20 seconds. When they hit 30 seconds, the node gets marked dead. If it helps, it seems like most of the warnings tend to appear when I’m moving data between nodes and cloud storage. To move data, I’m using multithreaded rsync; i.e. `gsutil -m rsync -r ...`. I’ve tried reducing the number of threads for rsync, but that hasn’t helped. Also, nodes are nowhere near max memory/CPU usage when the error occurs.

Any suggestions? Given that the error only occurs on 25% of runs and the logs seem to indicate that long timeouts are rare, I feel like increasing the heartbeat timeout would solve the issue, but I don’t see any options for that anywhere.

Thanks.

**Comment by rkooo567** [comment]
Maybe your rsync uses up all the network bandwidth which makes Ray heartbeat report slower? Is it possible to test with reduced data size when you move them around?

**Comment by ahah-figure** [comment]
The directories I'm rsync-ing contain somewhere between 10K and 50K chunks of data, each of which is around 20 MB. Happy to run a test and report back, but I think that reducing the number of data chunks or reducing the size of each chunk would just reduce the chance of a node's being marked dead, because rsync wouldn't run as long, right?

**Comment by richardliaw** [comment]
@ahah-figure can you try disabling syncing and trying again? https://docs.ray.io/en/master/tune/user-guide.html#distributed-checkpointing

**Comment by richardliaw** [comment]
Also, @ahah-figure do you mind sharing your cluster config?

cc @DmitriGekhtman and @ijrsvt

**Comment by ahah-figure** [comment]
@richardliaw the error happens when I'm not using tune as well. Normally, it occurs during initial processing, where I'm copying data from cloud storage to workers, but it also happens when I'm copying trained models from workers to cloud storage. Also, I am using the manual cluster setup outlined [here](https://docs.ray.io/en/releases-0.8.5/using-ray-on-a-cluster.html). I'm working on using the cluster launcher for GCP, but I have to make a few modifications to enable it to work with our security policies.

I've done some testing, and oddly, the issue only seems to arise on dataproc clusters with GPUs. The snippet below will cause the raylets to raise warnings about timeouts. I'm using python 3.8.10 and ray 1.3. The dataproc cluster has two nodes (n1-highmem-96) and 4 P4's per node. Please note that if I initialize the same cluster without GPUs, the issue doesn't arise.

```python
import os
import ray

SRC_DIR = "gs://path/with/a/lot/of/data"
DEST_DIR = "/tmp/tout"

@ray.remote(num_cpus=os.cpu_count())
def copy_data_to_worker():
    if os.path.isdir(DEST_DIR):
        os.system(f"rm -rf {DEST_DIR}")
    os.mkdir(DEST_DIR)
    cmd = f"gsutil -m rsync -r {SRC_DIR} {DEST_DIR}"
    os.system(cmd)


def main():
    ray.init(
        address="localhost:6379",
        _redis_password="None",
        ignore_reinit_error=True,
        include_dashboard=True,
    )
    num_nodes = len(ray.nodes())
    ray.get([copy_data_to_worker.remote() for _ in range(num_nodes)])


if __name__ == "__main__":
    main()
```

Above, `SRC_DIR` contains about 17K pickle files, each of which is 2 MB in size. They're smaller than I'd originally mentioned, because I'd hoped zipping them would alleviate the issue.

Let me know if I can provide any other info. Thanks!

**Comment by richardliaw** [comment]
Hmm, so it looks like there might be some OS scheduling issue here where the gsutil command hogs the threads and deprioritizes the Ray heartbeats. 

cc @ericl and @stephanie-wang , as this issue has shown up a couple times now.

So there is a parameter to increase the heartbeat timeout, but it's an internal config and not recommended to rely on: 

```
    - ray start
        --head
        --redis-port=6379
        --object-manager-port=8076
        --internal-config='{
            "num_heartbeats_timeout":300,
            "heartbeat_timeout_milliseconds":1000
        }'
```

**Comment by ahah-figure** [comment]
Thanks. Seems like changing the config values would certainly solve the issue in the short-run.

What are the default values for `num_heartbeats_timeout` and `heartbeat_timeout_milliseconds`? It seems like a node gets marked as dead if the heartbeat time is ever more than 30 seconds, so my guess would be the defaults are 

- `num_heartbeats_timeout` = 1
- `hearbeat_timeout_milliseconds` = 30000

Moving the zipped data to local takes less than five minutes, so as long as the time out is longer than that, everything should work fine.

**Comment by richardliaw** [comment]
Hey @ahah-figure -- https://github.com/ray-project/ray/blob/e925051ce412d60b74b142da14bea4a0a2cbf3ca/src/ray/common/ray_config_def.h#L46-L54 

close! It's the above.

**Comment by ahah-figure** [comment]
@richardliaw In bash, I tried to do `ray start --head --internal-config=...` flag, and I'm seeing 

`Error: no such option: --internal-config`

This happened with ray 1.3 and 1.4.1.

**Comment by rkooo567** [comment]
I think it is `--system-config`

**Comment by ahah-figure** [comment]
I used `--system-config={"num_heartbeats_timeout":300}` and ran the script I provided many times on many nodes, and I didn't get a single error.

Thanks everyone!

Would you like me to close this issue?

**Comment by richardliaw** [comment]
Hmm, unfortunately this is a real issue, but it's good to know this has a real workaround. Let's leave it open if possible :) 

Glad to hear that it was resolved though @ahah-figure !

**Comment by DmitriGekhtman** [comment]
Can we increase the default?
WDYT @rkooo567 ?

**Comment by richardliaw** [comment]
No, I think we shouldn't increase the default. 

Instead, we should just document this particular issue (and link it in the GCP stuff).

**Comment by rkooo567** [comment]
Yeah this is highly likely due to the fact that network bandwidth is taken by rsync. +1 on better documentation on various failure scenarios.

**Comment by venk-na** [comment]
Hi, The hearbeat signals from the nodes are sent via TCP or UDP?

**Comment by dayshah** [comment]
The health checks are sent via TCP, we use grpc to send them. If anyone is having issues with this now, the new health check env variables / configs are 
https://github.com/ray-project/ray/blob/45e369ad4d8baad1db1adac5feb6e2ea57a0d458/src/ray/common/ray_config_def.h#L876-L885


Bumping the health check period and timeout should help this, but if your Ray cluster is failing due to this either there's something up with your network or the GCS can't keep up with load due to the # of workers / actors / nodes.

**Comment by Hxinyue** [comment]
> I think it is `--system-config`

@rkooo567 When deploying a Ray cluster using the kuberay method and setting system-config under the rayStartParams section of the YAML file, you will encounter the error "Error: no such option".
